File: lib/services/geo_recommender.dart
// lib/services/geo_recommender.dart

import 'package:geolocator/geolocator.dart';
import 'package:http/http.dart' as http;
import 'package:flutter_dotenv/flutter_dotenv.dart';
import 'dart:convert';

String? _cachedPlaceType;
DateTime? _cacheTime;

/// 내부: Google Places API를 통해 장소 유형(한글)을 가져오는 함수
Future<String> _fetchPrimaryPlaceType(Position position) async {
  final apiKey = dotenv.env['GOOGLE_PLACES_API_KEY']!;
  final url = Uri.parse('https://places.googleapis.com/v1/places:searchNearby');

  final headers = {
    'Content-Type': 'application/json',
    'X-Goog-Api-Key': apiKey,
    'X-Goog-FieldMask': 'places.primaryTypeDisplayName.text',
  };

  final body = jsonEncode({
    'languageCode': 'ko',
    'maxResultCount': 1,
    'locationRestriction': {
      'circle': {
        'center': {
          'latitude': position.latitude,
          'longitude': position.longitude,
        },
        'radius': 100.0,
      }
    }
  });

  final resp = await http.post(url, headers: headers, body: body);
  if (resp.statusCode != 200) return '알 수 없음';
  final data = jsonDecode(resp.body);
  return (data['places']?[0]?['primaryTypeDisplayName']?['text'] as String?) ?? '알 수 없음';
}

/// 외부에서 위치(Position)를 받아 장소 유형을 반환 (캐시 사용 포함)
Future<String> getPlaceTypeFromPosition(Position position) async {
  // ① 캐시가 유효하면 반환
  if (_cachedPlaceType != null && _cacheTime != null) {
    final diff = DateTime.now().difference(_cacheTime!);
    if (diff.inMinutes < 10) {
      print("⚡ 캐시된 장소 유형 사용: $_cachedPlaceType");
      return _cachedPlaceType!;
    }
  }

  // ② 새로 호출
  final placeType = await _fetchPrimaryPlaceType(position);
  _cachedPlaceType = placeType;
  _cacheTime = DateTime.now();
  print("🆕 새 장소 유형 캐시됨: $placeType");
  return placeType;
}

/// 외부에서 호출: 장소 캐시 무효화
void invalidatePlaceTypeCache() {
  _cachedPlaceType = null;
  _cacheTime = null;
}

File: functions/src/geo_recommender.ts
// functions/src/geo_recommender.ts
import * as functions from "firebase-functions";
import * as admin from "firebase-admin";
import * as dotenv from "dotenv";
import fetch from "node-fetch";

dotenv.config();
const PLACES_API_URL = "https://places.googleapis.com/v1/places:searchNearby";
const API_KEY = process.env.GOOGLE_PLACES_API_KEY;

// Cloud Function: 장소 기반 북마크 추천
export const geoRecommender = functions.https.onCall(async (data, context) => {
  const { userId, latitude, longitude } = data;
    if (
    typeof userId !== "string" ||
    typeof latitude !== "number" ||
    typeof longitude !== "number"
    ) {
    throw new functions.https.HttpsError("invalid-argument", "userId (string), latitude (number), longitude (number)는 필수입니다.");
    }

  try {
    // 1) 장소 유형 조회
    const placeType = await fetchPrimaryPlaceType(latitude, longitude);

    // 2) logs에서 해당 장소 유형 + 태그 빈도 분석
    const logsSnapshot = await admin.firestore()
      .collection("logs")
      .where("userId", "==", userId)
      .where("location", "==", placeType)
      .get();

    const tagCount: Record<string, number> = {};
    logsSnapshot.forEach(doc => {
      const tags = (doc.get("tags") ?? []) as string[];
      tags.forEach(tag => {
        tagCount[tag] = (tagCount[tag] || 0) + 1;
      });
    });

    if (Object.keys(tagCount).length === 0) {
      return { placeType, recommendations: [] };
    }

    // 3) 가장 많이 사용된 태그 추출
    const topTag = Object.entries(tagCount).reduce((a, b) => a[1] > b[1] ? a : b)[0];

    // 4) 해당 태그 기반, 열람하지 않은 북마크 추천
    const bookmarksSnapshot = await admin.firestore()
      .collection("bookmarks")
      .where("userId", "==", userId)
      .where("tags", "array-contains", topTag)
      .where("wasOpened", "==", false)
      .get();

    const recommendations = bookmarksSnapshot.docs.map(doc => doc.get("title"));
    return { placeType, recommendations };

  } catch (err) {
    console.error("[geoRecommender] 오류:", err);
    throw new functions.https.HttpsError("internal", "추천 처리 중 오류가 발생했습니다.");
  }
});

// Places API 호출 → 장소 유형(한글명) 추출
async function fetchPrimaryPlaceType(lat: number, lng: number): Promise<string> {
  if (!API_KEY) throw new Error("Google Places API 키가 설정되지 않았습니다.");

  const res = await fetch(PLACES_API_URL, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-Goog-Api-Key": API_KEY,
      "X-Goog-FieldMask": "places.primaryTypeDisplayName.text",
    },
    body: JSON.stringify({
      languageCode: "ko",
      maxResultCount: 1,
      locationRestriction: {
        circle: {
          center: { latitude: lat, longitude: lng },
          radius: 50.0,
        },
      },
    }),
  });

  if (!res.ok) {
    console.error("[Places API] 응답 실패:", await res.text());
    return "알 수 없음";
    }
  const json = await res.json();
  return json?.places?.[0]?.primaryTypeDisplayName?.text || "알 수 없음";
}
